### DISCUSSION TITLE
SSH-Agent not working

### DISCUSSION THREAD

**Discussion by IamLunchbox** [i3/i3#5025] [Q&A] [answered]
Repost from an issue, since this issue is seemingly unrelated to i3 itself - but maybe someone has some advice:  
  
I recently migrated some virtual systems to Ubuntu 22.04. Since the upgrade necessary connection information for using the ssh-agent (`SSH_AUTH_SOCK` and `SSH_AGENT_PID`) is not written to the global environments anymore. So using `ssh-agent` leads to a well known "Could not open connection to your authentication agent" error.

I've read through countless stackoverflow questions, which only offer a solution to open the agent in a terminal session using `eval $(ssh-agent)`. But this method a) spawns several agents and b) is not working, if another process needs to use the key - a code editor for example.

I even tweaked my i3-config with the following: `exec --no-startup-id /usr/bin/gnome-keyring-daemon --start --components=ssh,secrets,pkcs11`, but neither this nor making sure `use-ssh-agent` is in /etc/X11/Xsession.options helped

Since everything worked fine in Ubuntu 20.04 i figured, that it could be a problem with a newer i3-version or another related package? Logging in with LightDM and GDM3 into Gnome works fine. I noticed a change of the value-format for SSH_AUTH_SOCK from `/tmp/ssh-xxxxx/agent.pid` to `/run/user/1000/keyring/ssh`. Maybe this is the culprit? This variable is in the given Gnome enviroment, but disappears when I use i3.
I am using:  
- i3 version: 4.20.1
- Ubuntu 22.04
- configuration is available in my [dotfiles](https://raw.githubusercontent.com/IamLunchbox/dotfiles/main/config/desktop/config/i3/config). 

Logs:  
I went through the debug-log, but scrolling and grep'ing through it brought no hints of any misconfiguration regarding ssh..

Did someone stumble upon the same issue? Is it related to i3, X or maybe a revamped gpg-component?

I'd be grateful for any advice :)

**Comment by alexanderustinov** [comment]
Adding
```
case "$DESKTOP_SESSION" in
    i3)
       	eval $(ssh-agent)
        ;;
    sway)
       	eval $(ssh-agent)
        ;;
esac
```
to `~/.profile` followed by re-login works for me with i3-4.20.1.

**Comment by IamLunchbox** [comment, marked as answer]
Hey Alex,
I was able to find a specific bug report for X11, which was responsible for this behaviour [here](https://bugs.launchpad.net/ubuntu/+source/xorg/+bug/1922414). This issue primarily applies to X11-based launchers in conjunction with Ubuntu 21.10 and 22.04.

A temporary solution, which was given within the linked thread above, worked for me:
`cat /etc/X11/Xsession.d/01x11-has_option`
```sh
# This file is sourced by Xsession(5), not executed.
#
# workaround created by Settel
# defines "has_option" unless it is already defined
# see https://bugs.launchpad.net/ubuntu/+source/xorg/+bug/1922414
#


if type -t has_option >/dev/null; then
  return
fi


OPTIONS="$(
  if [ -r "$OPTIONFILE" ]; then
    cat "$OPTIONFILE"
  fi
  if [ -d /etc/X11/Xsession.options.d ]; then
    run-parts --list --regex '\.conf$' /etc/X11/Xsession.options.d | xargs -d '\n' cat
  fi
)"

has_option() {
  # Ensure that a later no-foo overrides an earlier foo
  if [ "$(echo "$OPTIONS" | grep -Eo "^(no-)?$1\>" | tail -n 1)" = "$1" ]; then
    return 0
  else
    return 1
  fi
}
```
